### PR TITLE
Replace Common protocol with SecureChannel

### LINKED ISSUE #5379: Resolve Message and Protocol IDs for Common/SecureChannel Protocols
#### Problem
 "Common Protocol" is currently being used synonymously with the Secure Channel Protocol, even though nothing is currently defined to be part of the "Common Protocol". There should be no references to "Common Protocol" in code.

 #### Proposed Solution
Change all code references to the Common Protocol to instead be part of the Secure Channel Protocol. This includes:
- the StatusCode and MsgType enums in [src/protocols/common/Constants.h](https://github.com/project-chip/connectedhomeip/blob/master/src/protocols/common/Constants.h).
- the Protocol IDs defined in [src/Protocols.h](https://github.com/project-chip/connectedhomeip/blob/master/src/protocols/Protocols.h)
- `StatusReport` needs to be added to the `MsgType` enum in [src/protocols/secure_channel/Constants.h](https://github.com/project-chip/connectedhomeip/blob/master/src/protocols/secure_channel/Constants.h).
- change any other code references to Common Protocol ID to Secure Channel

### PR DESCRIPTION
**Pull request by holbrookt**
#### Problem
"Common" protocol was being used synonymously with SecureChannel, even though the spec does not currently define anything to exist in a Common protocol.

Some StatusReport code is currently defined as part of Common protocol when it should be SecureChannel.

This is closely related to #5359 

 #### Summary of Changes
    - remove references to Common protocol and replace with SecureChannel
    - change SecureChannel protocol ID to 0
    - move StatusCode definitions to
    src/protocols/secure_channel/Constants.h
    - initialize PayloadHeader with kProtocol_NotSpecified instead of 0

 Fixes #5379

### CODE REVIEW

File: src/app/CommandSender.cpp [hunk 1]
```diff
@@ -145,7 +145,7 @@ CHIP_ERROR CommandSender::ProcessCommandDataElement(CommandDataElement::Parser &
             err = CHIP_NO_ERROR;
             ChipLogDetail(DataManagement, "Add Status code for empty command, cluster Id is %d", clusterId);
             // Todo: Define protocol code for StatusCode
-            AddStatusCode(0, chip::Protocols::kProtocol_Protocol_Common, 0, clusterId);
+            AddStatusCode(0, chip::Protocols::kProtocol_SecureChannel, 0, clusterId);
```
Comment (bzbarsky-apple) [inline src/app/CommandSender.cpp:148]: Preexisting, but `COMMON_STATUS_SUCCESS` for the first arg?

And perhaps file a followup to have enum classes, one per protocol, for the standard-protocol status codes?

File: src/protocols/Protocols.h [hunk 2]
```diff
@@ -39,8 +39,7 @@ enum CHIPProtocolId
     //
     // NOTE: Do not attempt to allocate these values yourself.
 
-    kProtocol_Protocol_Common     = (kChipVendor_Common << 16) | 0x0000, // Common Protocol
-    kProtocol_SecureChannel       = (kChipVendor_Common << 16) | 0x0001, // Secure Channel Protocol
+    kProtocol_SecureChannel       = (kChipVendor_Common << 16) | 0x0000, // Secure Channel Protocol
     kProtocol_Echo                = (kChipVendor_Common << 16) | 0x0002, // Echo Protocol
```
Comment (bzbarsky-apple) [inline src/protocols/Protocols.h:43]: We're going to fix the rest of these protocol numbers too, to match the spec.  Please file a followup issue?

File: src/protocols/bdx/BdxTransferSession.cpp [hunk 3]
```diff
@@ -431,8 +431,8 @@ CHIP_ERROR TransferSession::HandleMessageReceived(System::PacketBufferHandle msg
 
         mTimeoutStartTimeMs = curTimeMs;
     }
-    else if (payloadHeader.GetProtocolID() == Protocols::kProtocol_Protocol_Common &&
-             payloadHeader.GetMessageType() == static_cast<uint8_t>(Protocols::Common::MsgType::StatusReport))
+    else if (payloadHeader.GetProtocolID() == Protocols::kProtocol_SecureChannel &&
+             payloadHeader.GetMessageType() == static_cast<uint8_t>(Protocols::SecureChannel::MsgType::StatusReport))
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/BdxTransferSession.cpp:435]:
```suggestion
    else if (payloadHeader.HasMessageType(Protocols::SecureChannel::MsgType::StatusReport))
```

File: src/protocols/bdx/tests/TestBdxTransferSession.cpp [hunk 4]
```diff
@@ -127,8 +127,9 @@ void VerifyStatusReport(nlTestSuite * inSuite, void * inContext, const System::P
 
     err = payloadHeader.Decode(msg->Start(), msg->DataLength(), &headerSize);
     NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
-    NL_TEST_ASSERT(inSuite, payloadHeader.GetProtocolID() == Protocols::kProtocol_Protocol_Common);
-    NL_TEST_ASSERT(inSuite, payloadHeader.GetMessageType() == static_cast<uint8_t>(Protocols::Common::MsgType::StatusReport));
+    NL_TEST_ASSERT(inSuite, payloadHeader.GetProtocolID() == Protocols::kProtocol_SecureChannel);
+    NL_TEST_ASSERT(inSuite,
+                   payloadHeader.GetMessageType() == static_cast<uint8_t>(Protocols::SecureChannel::MsgType::StatusReport));
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/tests/TestBdxTransferSession.cpp:132]:
```suggestion
    NL_TEST_ASSERT(inSuite, payloadHeader.HasMessageType(Protocols::SecureChannel::MsgType::StatusReport));
```

File: src/transport/SecureSessionMgr.cpp [hunk 5]
```diff
@@ -161,14 +161,13 @@ CHIP_ERROR SecureSessionMgr::SendMessage(SecureSessionHandle session, PayloadHea
     VerifyOrExit(admin != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
     localNodeId = admin->GetNodeId();
 
-    if (payloadHeader.HasMessageType(Protocols::SecureChannel::MsgType::MsgCounterSyncReq) ||
```
Comment (bzbarsky-apple) [inline src/transport/SecureSessionMgr.cpp:164]: Why the changes here?  They don't seem related to this PR......

**Review by andy31415 — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

File: src/app/CommandSender.cpp (hunk 1 above)
Comment (holbrookt) [reply src/app/CommandSender.cpp:148]: instead of `COMMON_STATUS_SUCCESS` it should actually be `chip::Protocols::SecureChannel::GeneralStatusCode::kSuccess` (that's a mouthful...).

Will file a followup about protocol codes

File: src/protocols/Protocols.h (hunk 2 above)
Comment (holbrookt) [reply src/protocols/Protocols.h:43]: filed https://github.com/project-chip/connectedhomeip/issues/5441

File: src/app/CommandSender.cpp (hunk 1 above)
Comment (holbrookt) [reply src/app/CommandSender.cpp:148]: Discussed offline, general codes are already defined in this PR in src/protocols/secure_channel/Constants.h

Comment (bzbarsky-apple) [reply src/app/CommandSender.cpp:148]: I was confused about what this line does.  The thing that would make sense is to have per-protocol enum classes for the protocol-specific codes, and then perhaps a version of AddStatusCode that derives the protocol from the protocol-specific code, if there is one.  But that couldn't be used in this case, since there is no protocol-specific code here.

File: src/transport/SecureSessionMgr.cpp (hunk 5 above)
Comment (holbrookt) [reply src/transport/SecureSessionMgr.cpp:164]: This PR changes the SecureChannel protocol ID to 0, and this exposed a bug here. When calling `SendEncryptedMessage()` (defined above the function shown here), it default-constructs a `PayloadHeader`, which results in a `PayloadHeader` with `ProtocolID = 0` (SecureChannel) and `MessageID = 0` (MsgCounterSyncReq) and this was unintentionally causing the condition here to be true.

Comment (holbrookt) [reply src/transport/SecureSessionMgr.cpp:164]: I've requested @yufengwangca to review this because he recently edited this code.

File: src/protocols/bdx/BdxTransferSession.cpp (hunk 3 above)
Comment (holbrookt) [reply src/protocols/bdx/BdxTransferSession.cpp:435]: Good catch, thanks.

File: src/protocols/bdx/tests/TestBdxTransferSession.cpp (hunk 4 above)
Comment (holbrookt) [reply src/protocols/bdx/tests/TestBdxTransferSession.cpp:132]: Done

File: src/transport/SecureSessionMgr.cpp (hunk 5 above)
Comment (bzbarsky-apple) [reply src/transport/SecureSessionMgr.cpp:164]: > which results in a PayloadHeader with ProtocolID = 0 (SecureChannel) and MessageID = 0 (MsgCounterSyncReq) 

Uh... we should probably have the default constructor use an invalid protocol id, or message id, or both!

Comment (holbrookt) [reply src/transport/SecureSessionMgr.cpp:164]: That's a great idea! I can add that to this PR if you think it's relevant enough

Comment (yufengwangca) [reply src/transport/SecureSessionMgr.cpp:164]: Should we really want to use 0 as one of valid protocol ID?  maybe we should modify spec to avoid similar occident.

Comment (bzbarsky-apple) [reply src/transport/SecureSessionMgr.cpp:164]: We could try to change the spec, sure.  Might be easier to change the code, though.

@holbrookt Followup for the constructor change is ok, assuming @yufengwangca is OK with the change here.

Comment (bzbarsky-apple) [reply src/transport/SecureSessionMgr.cpp:164]: The other thing we could do is not pass random default-constructed things to functions that aren't supposed to use them and use Optional<> as needed instead.  Or a better API that doesn't involve constructing "headers" that are then unused.

Comment (yufengwangca) [reply src/transport/SecureSessionMgr.cpp:164]: I am ok to solve this issue by putting default value in constructor for now. We already have default value kProtocol_NotSpecified defined.

**Comment by holbrookt** [comment]
@bzbarsky-apple @yufengwangca In the most recent commit I add the default-initialization of PayloadHeader with `kProtocol_NotSpecified`. I also tried this change with the original code from SecureSessionMgr.cpp and a bunch of unit tests failed. Which means sometimes 
```
if (payloadHeader.HasMessageType(Protocols::SecureChannel::MsgType::MsgCounterSyncReq)
```

is probably still evaluating as true unintentionally sometimes

File: src/transport/raw/tests/TestMessageHeader.cpp [hunk 6]
```diff
@@ -50,7 +50,7 @@ void TestPayloadHeaderInitialState(nlTestSuite * inSuite, void * inContext)
 
     NL_TEST_ASSERT(inSuite, header.GetMessageType() == 0);
     NL_TEST_ASSERT(inSuite, header.GetExchangeID() == 0);
-    NL_TEST_ASSERT(inSuite, header.GetProtocolID() == 0);
+    NL_TEST_ASSERT(inSuite, header.GetProtocolID() == static_cast<uint16_t>(Protocols::kProtocol_NotSpecified));
```
Comment (bzbarsky-apple) [inline src/transport/raw/tests/TestMessageHeader.cpp:53]: This cast doesn't seem valid: `kProtocol_NotSpecified` is not in fact a value that fits in 16 bits....

**Comment by bzbarsky-apple** [comment]
> and a bunch of unit tests failed

Which ones?  I just tries this locally and it seemed to work, but it's possible that the tests involved are disabled on my platform...

File: src/transport/raw/tests/TestMessageHeader.cpp (hunk 6 above)
Comment (holbrookt) [reply src/transport/raw/tests/TestMessageHeader.cpp:53]: Seems like this is a bug in the PayloadHeader API. It stores ProtocolID as a 16 bit value when it should be 32 according to spec. That seems like it could be out of scope of this PR though. I could file an issue and follow up PR instead. what do you think?

**Comment by holbrookt** [comment]
> > and a bunch of unit tests failed
> 
> Which ones? I just tries this locally and it seemed to work, but it's possible that the tests involved are disabled on my platform...

I just tried it again and they passed... It's possible I just didn't save my changes or maybe I should have done a clean build first. I'll upload another commit resetting the changes for that file

File: src/transport/raw/tests/TestMessageHeader.cpp (hunk 6 above)
Comment (bzbarsky-apple) [reply src/transport/raw/tests/TestMessageHeader.cpp:53]: PayloadGeader stores an optional 16-bit vendor id and a 16-bit protocol id.  This is also what the spec defines for the payload header...  How this interacts with the concept of "protocol id" in the rest of the code, which is a combination of those two values, is a good question, and I agree that it's out of scope for this PR.  

We do have a bunch of code that tests the protocol id without considering the vendor id.  That code is probably all broken if someone sends a vendor-scoped protocol id.  Again, seems out of scope for this PR.

For this PR, maybe just stick to using 0xFFFF for the default value for now and file followups to make this stuff more consistent across the spec and less footgunny?

Comment (holbrookt) [reply src/transport/raw/tests/TestMessageHeader.cpp:53]: Thanks, I was reading the PayloadHeader too fast and missed that it splits VendorID out.

For this code and the initializer, I can mask out the lower 16 bits explicitly.

**Review by bzbarsky-apple — APPROVED** [review]
